### PR TITLE
methods used on in-place upgrades

### PR DESCRIPTION
**Pull request by paulomach**
## Issue

New lib methods for in-place upgrades implemented for k8s not yet in lib.

## Solution

* Bring methods from k8s [PR#255](https://github.com/canonical/mysql-k8s-operator/pull/255)
* bit of increase on test coverage

### CODE REVIEW

File: lib/charms/mysql/v0/mysql.py [hunk 1]
```diff
@@ -123,6 +123,15 @@ def wait_until_mysql_connection(self) -> None:
 class Error(Exception):
```
Comment (shayancanonical) [inline lib/charms/mysql/v0/mysql.py:123]: may need to update libpatch

Comment (paulomach) [reply lib/charms/mysql/v0/mysql.py:123]: 518b8c9 :man_facepalming:

**Review by shayancanonical — DISMISSED** [review]

File: lib/charms/mysql/v0/mysql.py [hunk 2]
```diff
@@ -123,6 +123,15 @@ def wait_until_mysql_connection(self) -> None:
...
     """Base class for exceptions in this module."""
 
+    def __init__(self, message: Optional[str] = None) -> None:
```
[the remark this answers is not in the document]
Comment (paulomach) [reply lib/charms/mysql/v0/mysql.py]: a9b14f8

File: lib/charms/mysql/v0/mysql.py [hunk 3]
```diff
@@ -975,6 +996,32 @@ def remove_router_from_cluster_metadata(self, router_id: str) -> None:
             logger.exception(f"Failed to remove router from metadata with ID {router_id}")
             raise MySQLRemoveRouterFromMetadataError(e.message)
 
+    def set_dynamic_variable(
+        self, variable: str, value: str, persist: bool = False, instance: Optional[str] = None
+    ) -> None:
+        """Set a dynamic variable value for the instance.
```
[the remark this answers is not in the document]
Comment (paulomach) [reply lib/charms/mysql/v0/mysql.py:1006]: a9b14f8

File: lib/charms/mysql/v0/mysql.py [hunk 4]
```diff
@@ -975,6 +996,32 @@ def remove_router_from_cluster_metadata(self, router_id: str) -> None:
...
+            value: The value to set the variable to
+            persist: Whether to persist the variable value across restarts
+            instance: instance address to run on, default to current
+
+        Raises:
+            MySQLSetVariableError
+        """
+        logger.debug(f"Setting {variable} to {value} on {instance or self.instance_address}")
+        set_var_command = [
+            f"shell.connect('{self.server_config_user}:{self.server_config_password}@{instance or self.instance_address}')",
```
[the remark this answers is not in the document]
Comment (paulomach) [reply lib/charms/mysql/v0/mysql.py]: a9b14f8

File: lib/charms/mysql/v0/mysql.py [hunk 5]
```diff
@@ -1579,6 +1629,34 @@ def get_cluster_primary_address(
...
+            return None
+        for label, value in status["defaultreplicaset"]["topology"].items():
+            if value["memberrole"] == "primary":
+                return label
+
+    def set_cluster_primary(self, primary_address: str) -> None:
+        """Set the cluster primary.
+
+        Args:
+            primary_address: The address of the cluster's primary
```
[the remark this answers is not in the document]
Comment (paulomach) [reply lib/charms/mysql/v0/mysql.py]: a9b14f8

File: lib/charms/mysql/v0/mysql.py [hunk 6]
```diff
@@ -1605,6 +1683,28 @@ def get_cluster_members_addresses(self) -> Optional[Iterable[str]]:
...
+            "    print('SAME_VERSION')",
+        ]
+
+        try:
+            output = self._run_mysqlsh_script("\n".join(check_command))
+            if "SAME_VERSION" in output:
+                return
+            result = json.loads(output)
+            if result["errorCount"] == 0:
+                return
```
[the remark this answers is not in the document]
Comment (paulomach) [reply lib/charms/mysql/v0/mysql.py:1716]: a9b14f8

File: lib/charms/mysql/v0/mysql.py [hunk 7]
```diff
@@ -123,6 +123,15 @@ def wait_until_mysql_connection(self) -> None:
...
     """Base class for exceptions in this module."""
 
+    def __init__(self, message: Optional[str] = None) -> None:
+        """Initialize the Error class.
+
+        Args:
+            message: Optional message to pass to the exception.
+        """
+        super().__init__(message)
+        self.message = message or ""
```
[the remark this answers is not in the document]
Comment (paulomach) [reply lib/charms/mysql/v0/mysql.py]: not sure I understood it

File: tests/unit/test_mysql.py [hunk 8]
```diff
@@ -1366,7 +1369,10 @@ def test_empty_data_files(
             group="test-group",
         )
 
-        _expected_commands = "find mysql/data/directory -not -path mysql/data/directory/#mysql_sst_* -not -path mysql/data/directory -delete".split()
+        _expected_commands = (
+            "find mysql/data/directory -not -path "
+            "mysql/data/directory/#mysql_sst_* -not -path mysql/data/directory -delete"
+        ).split()
```
[the remark this answers is not in the document]
Comment (paulomach) [reply tests/unit/test_mysql.py]: a9b14f8

**Review by shayancanonical — APPROVED** [review]

**Pull request merged**